src/main/java/org/quiltmc/loader/api/QuiltLoader.java

```diff
@@ -0,0 +1,191 @@
+
+/**
+ * The public-facing FabricLoader instance.
+ *
+ * <p>To obtain a working instance, call {@link #getInstance()}.</p>
+ *
+ * @since 0.4.0
+ */
+public interface QuiltLoader {
+	/**
+	 * Returns the public-facing Quilt Loader instance.
+	 */
+	static QuiltLoader getInstance() {
+		if (QuiltLoaderImpl.INSTANCE == null) {
+			throw new RuntimeException("Accessed QuiltLoader too early!");
+		}
+
+		return QuiltLoaderImpl.INSTANCE;
+	}
+
+	/**
+	 * Returns all entrypoints declared under a {@code key}, assuming they are of a specific type.
+	 *
+	 * @param key  the key in entrypoint declaration in {@code fabric.mod.json}
+	 * @param type the type of entrypoints
+	 * @param <T>  the type of entrypoints
+	 * @return the obtained entrypoints
+	 * @see #getEntrypointContainers(String, Class)
+	 */
+	<T> List<T> getEntrypoints(String key, Class<T> type);
+
+	/**
+	 * Returns all entrypoints declared under a {@code key}, assuming they are of a specific type.
+	 *
+	 * <p>The entrypoint is declared in the {@code fabric.mod.json} as following:
+	 * <pre><blockquote>
+	 *   "entrypoints": {
+	 *     "&lt;a key&gt;": [
+	 *       &lt;a list of entrypoint declarations&gt;
+	 *     ]
+	 *   }
+	 * </blockquote></pre>
+	 * Multiple keys can be present in the {@code entrypoints} section.</p>
+	 *
+	 * <p>An entrypoint declaration indicates that an arbitrary notation is sent
+	 * to a {@link LanguageAdapter} to offer an instance of entrypoint. It is
+	 * either a string, or an object. An object declaration
+	 * is of this form:<pre><blockquote>
+	 *   {
+	 *     "adapter": "&lt;a custom adatper&gt;"
+	 *     "value": "&lt;an arbitrary notation&gt;"
+	 *   }
+	 * </blockquote></pre>
+	 * A string declaration {@code <an arbitrary notation>} is equivalent to
+	 * <pre><blockquote>
+	 *   {
+	 *     "adapter": "default"
+	 *     "value": "&lt;an arbitrary notation&gt;"
+	 *   }
+	 * </blockquote></pre>
+	 * where the {@code default} adapter is the {@linkplain LanguageAdapter adapter}
+	 * offered by Fabric Loader. </p>
+	 *
+	 * @param key  the key in entrypoint declaration in {@code fabric.mod.json}
+	 * @param type the type of entrypoints
+	 * @param <T>  the type of entrypoints
+	 * @return the entrypoint containers related to this key
+	 * @throws EntrypointException if a problem arises during entrypoint creation
+	 * @see LanguageAdapter
+	 */
+	<T> List<EntrypointContainer<T>> getEntrypointContainers(String key, Class<T> type);
+
+	/**
+	 * Get the current mapping resolver.
+	 *
+	 * <p>When performing reflection, a mod should always query the mapping resolver for
+	 * the remapped names of members than relying on other heuristics.</p>
+	 *
+	 * @return the current mapping resolver instance
+	 * @since 0.4.1
+	 */
+	MappingResolver getMappingResolver();
+
+	/**
+	 * Gets the container for a given mod.
+	 *
+	 * @param id the ID of the mod
+	 * @return the mod container, if present
+	 */
+	Optional<ModContainer> getModContainer(String id);
```

Comment:
Do we want to keep this as optional or change to nullable?

Reply:
`@Nullable` feels lazy, but `Optional` adds extra verbosity and time. I think the choice should be more decided on QuiltLoader acting as more functional or imperative programming